### PR TITLE
Connection Form Refactor - Connection Form Service

### LINKED ISSUE #15603: Create ConnectionFormService and migrate Connection workflow to use it
Currently, there are a number of methods for getting, changing, and sending the connection object during connection creation and editing.  The goal is to get all of that logic within one Service that can be utilized from within the Connection Form.

Create a `ConnectionFormService` that returns methods for and migrate existing calls to use it.

This should be sure to include:

Creating:
- Generating the form’s initial values from specs and source schema
- After form submit, taking the form values and creating the connection (sending it to the database)

Editing:
- Generating the form’s initial values from specs, and existing connection object
- Refreshing the source schema (this should be able to be called from the button directly and then override the form’s sync catalog values with the found sync catalog)
- Updating the saved connection based on the received form values

This Service should be well tested. Depending on the result of the PR, set up a test-it days session.

### PR DESCRIPTION
**Pull request by krishnaglick**
Closes #15603

## What
The Connection Create/Edit code is a dense and complex jungle. This is first portion of work to simplify it, and increase maintainability.

#16293 and #16303 spun out of this work as ways to improve the reliability and readability of existing code.

## Recommended reading order
Start with:
- airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
- airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
There was a lot of logic wrapped up in the ConnectionForm. It contains the Formik form, handled passing the dirty state up VIA an ugly chain of logic, coordinated the submit & afterSubmit logic, and a few other misc things. Plus it handled various bits of prop drilling. A lot of this was moved into the ConnectionFormService which keeps us from prop drilling, and separates helps keep the render logic separate.

Then:
- airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
- airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx

These two components will likely see greater change in the next PR in this series. There was a lot of back-and-forth with the ConnectionForm:
- CreateConnectionContent prop-drilled the `onSelectFrequency` analytics call
- CreateConnectionContent's onSubmitConnectionStep returned a method to be called `afterSubmit` in what I felt was an awkward way. This has been streamlined.
- ReplicationView previously had an awkward, upwards chained bit of tracking to determine if a form is dirty. Now it just utilizes the global form dirty tracking and all that messy logic has been removed.
There were a few other changes, like the onSubmit returning `{ submitCancelled: true }` now utilizes its try/catch, and both components get to do less prop drilling.

After #16293 and #16303 are merged this diff will look cleaner, but there's various components like `StreamHeader.tsx` where instead of getting a value from its props, it's just pulling it out of the context.
Lastly there's some testing that's worth looking over!

One final note is `OperationsSection.tsx` had a Wrapper prop passed in. That's been cleaned up to just import the appropriate wrapping components.

## 🚨 User Impact 🚨
This work touches the major pieces of our Connection Create and Edit pages. They're quite touchy and core to the application so this _potentially_ has a large user impact if bugs are introduced so please test carefully!

### CODE REVIEW

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:95]: These are in here to prevent prop-drilling.

**Comment by tealjulia** [comment]
Looks like the field for Connection Name input disappeared during connection creation.

**Comment by krishnaglick** [comment]
> Looks like the field for Connection Name input disappeared during connection creation.

Should be fixed!

File: airbyte-webapp/src/hooks/services/useSourceHook.tsx
Comment (edmundito) [inline airbyte-webapp/src/hooks/services/useSourceHook.tsx:187]: How come this went from async to not async?

**Review by edmundito — COMMENTED** [review]
Just had a few comments

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:27]: This is copied from now-old code.
https://github.com/airbytehq/airbyte/blob/65584ca22435581d621b3abd1ad31366a72e706c/airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx#L44
But we take a partial connection when making a new one and this lets us require some specific fields while maintaining types.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:158]: I feel like this is a little bit better of a method than sometimes returning submitCancelled. Happy to discuss.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:219]: `const { mode } = useConnectionFormService();`
It's just moved into the service. Did my best to clean up all the prop drilling.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/OperationsSection.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/OperationsSection.tsx:37]: That was fun to figure out: it was done so that _if_ this should render then we have the correct styles.
This is a little less janky I think :D

File: airbyte-webapp/src/hooks/services/useSourceHook.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/useSourceHook.tsx:187]: The async wasn't affecting anything. Awaiting this within a function call that itself was not awaited meant it wasn't actually blocking anything.

File: airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx
Comment (tealjulia) [reply airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx:68]: Does it make sense to make this part of the after-submit _always_?  When you edit a connection on the replication view, it often (always?) triggers a reset.  It could make sense that submitting the form takes you to `/connection/:id` where you can see the status of that job.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (tealjulia) [inline airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:1]: Is it just the two lines where we weren't able to get away without the `any`?  Could we just disable eslint for those lines?

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (tealjulia) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:186]: Is this `TODO` in this PR or a future one?

File: airbyte-webapp/src/pages/ConnectionPage/pages/CreationFormPage/CreationFormPage.tsx
Comment (tealjulia) [inline airbyte-webapp/src/pages/ConnectionPage/pages/CreationFormPage/CreationFormPage.tsx:163]: 👋

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (edmundito) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:27]: Might be worth extracting this into a type so that it's clear.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (edmundito) [inline airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:48]: I'd like to see more testing here. This only tests for onSubmit, but there are 11 other fields in the service

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (edmundito) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:186]: ?

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (edmundito) [inline airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:86]: I think it's odd to expose the observable in the hook. Preferably the hook should provide some interface that simplifies the flow and internally use the observable to track.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:1]: Those 6 lines yeah. But I feel way less strongly about using `any` in a test file over real code. I could move the ignore down if you'd like!

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx]: Done!

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:186]: For next PR.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx]: Removed!

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:48]: This is mostly just utilizing other hooks and passing those values along. I'm happy to discuss but I feel like testing those means writing tests for just those hooks, not this service.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:219]: @krishnaglick it looks like `<TreeViewContainer />` is a styled component that doesn't do anything with the mode.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (tealjulia) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:48]: Are those hooks tested?  I think we want to make sure this entire workflow has good coverage.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:219]: Oh! Got it. ty

**Review by teallarson — COMMENTED** [review]
A few comments looking through this evening and testing it out locally!  Some of the comments are existing bugs that we may want to address in the upcoming Issue rather than this one.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.test.tsx:48]: Added tests to most of these hooks.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx
Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:218]: This is broken on `master`, not related to this PR!

Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:218]: Disregard.  This was never broken, I just did not understand that you have to click "apply" for this feature to work.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (tealjulia) [reply airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:43]: Moving to connection form refactor, this is present on `master` as well.

**Comment by krishnaglick** [comment]
I've pulled out some portions of the code into #16303 and #16293. Those two, when merged, will generate some minor conflicts but will clean up this PR.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/components/SyncCatalogField.tsx:218]: I made the same mistake 😞

File: airbyte-webapp/src/components/CreateConnectionContent/components/TryAfterErrorBlock.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/components/CreateConnectionContent/components/TryAfterErrorBlock.tsx:22]: I believe I forgot to remove this as part of the side-PR's. This was unused.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/SettingsView.test.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/SettingsView.test.tsx:24]: Had to make this change to prevent this test from failing since `AnalyticsProvider` was getting overridden.

File: airbyte-webapp/src/test-utils/testutils.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/test-utils/testutils.tsx:45]: This was added and removed due to me misunderstanding a failing test. But it needs to be here for my tests.

File: airbyte-webapp/src/utils/errorStatusMessage.test.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/utils/errorStatusMessage.test.tsx:3]: Missed renaming this from one of the side-PRs

File: airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: Adding visibility to this question

Comment (tealjulia) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: Tagging @nataliekwong on that one!

Currently we only trigger the Frequency Select analytics call when creating a connection. The alternative here is we could send analytics whenever a user selects a frequency (ie: creating or editing).

If we do change it, I imagine we will also want to track the mode (create or edit) in the call name.

**Comment by tealjulia** [comment]
Code looks good to me, I feel good about merging this if the test-it day findings are solid, which they seem to be so far.

File: airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx
Comment (nataliekwong) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: Yep, agreed with @tealjulia 's comment on splitting the call into create and edit separately.

Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: Would you be okay with me just including the `mode` in the call? As part of the Event?

```
        analyticsService.track(Namespace.CONNECTION, Action.FREQUENCY, {
          actionDescription: "Frequency selected",
          frequency: item.label,
          connector_source_definition: connection.source.sourceName,
          connector_source_definition_id: connection.source.sourceDefinitionId,
          connector_destination_definition: connection.destination.destinationName,
          connector_destination_definition_id: connection.destination.destinationDefinitionId,
          available_streams: connection.syncCatalog.streams.length,
          enabled_streams: enabledStreams,
          mode: mode,
        });
        ```

Comment (nataliekwong) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: Yes, it's fine to use that. Could we rename it from "mode" to "type"? Are the values for it going to be "new" and "existing"? Or "create" and "update"?

And just to double check, the name the call is going to remain as "airbyte_ui_cloud.airbyte_ui_frequency_set"? (Not frequency selected)

Comment (krishnaglick) [reply airbyte-webapp/src/views/Connection/ConnectionForm/ConnectionForm.tsx:211]: The name should remain the same. I'll pass `type` instead of mode. The value will be one of these: `"create" | "edit" | "readonly";`
I could transcribe create to "new" and edit to "existing". Technically readonly should never be sent.

**Review by teallarson — APPROVED** [review]
Test it Day Doc looks good.  Testing and code look good.

This should have two PR approvals before merge though.

File: airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx
Comment (edmundito) [inline airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx:70]: This looks too high in the component tree and should be called from the Connection form. The reason is that the ConnectionForm sets up the ConnectionChangeTracker, and should be responsible for clearing it after submission. In this case, this component, which has no idea what the form does, assumes that the tracker is being used.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (edmundito) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:169]: Throwing should only be used when there is a misuse of the function or an error. In this case, this is a valid action that the user can take. This forces the developer using to catch errors and remember to check that one of them is actually a cancel action.

**Review by edmundito — COMMENTED** [review]
Finally got around to doing another pass.

File: airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx
Comment (tealjulia) [reply airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx:70]: The next PR brings the FormikContext "out" a layer at which point this would be the right place for this call.  Does it seem reasonable to leave this for now knowing it will be much more readily adressable in the next PR?

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:169]: This particular bit of code is already encapsulated in a try/catch due to its async nature.
I initially made this change as I made the function return a new connection which felt awkward alongside this. This now doesn't return anything so I could revert it however as part of the next portion of this it should see further refactoring and I believe I will be able to clean up this usage better then.

**Comment by lmossman** [comment]
@krishnaglick I'm getting a build error when running this locally now, same one as in CI: https://github.com/airbytehq/airbyte/actions/runs/3047952989/jobs/4912510015#step:7:144
I'm gonna update the branch to pull in the most recent changes from master so that I can run this locally, hopefully that is fine!

File: airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx:39]: This all will get pulled out in part 2.

File: airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/components/CreateConnectionContent/CreateConnectionContent.tsx:72]: This will be changed in part 2.

File: airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/hooks/services/Connection/ConnectionFormService.tsx:82]: This will be changed in part 2.

File: airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx
Comment (krishnaglick) [reply airbyte-webapp/src/pages/ConnectionPage/pages/ConnectionItemPage/components/ReplicationView.tsx:169]: This will be changed in part 2.

File: airbyte-webapp/src/views/Connection/ConnectionForm/components/ScheduleField.tsx
Comment (krishnaglick) [inline airbyte-webapp/src/views/Connection/ConnectionForm/components/ScheduleField.tsx:58]: Declared at the appropriate level now.

**Review by edmundito — APPROVED** [review]

**Comment by krishnaglick** [comment]
Thanks for the all-clear Edmundo!
I'm going to merge this Monday, don't want to blow up on-call _just in case_

**Pull request merged**
